VIMC-3503: Add usethis-style orderly_use_resource

richfitz opened the pull request:
See #171 where @jeffeaton asks helper functions.  This is something also discussed in #160 (VIMC-3404).  Resources is probably the easiest case because it's an easy part of the yaml to edit (it's just an array) and because the files are already there. Unlike the requested spec, this PR does not copy the file in, though we could look to add that later perhaps.

Editing yaml while preserving non-syntactic properties is hard in general, so this PR implements some basic heuristics to hopefully do an ok job of it.  It will not work well on pathalogical yaml, but should do ok on the likely simple yaml that people actually write.

Fixes #171

richfitz commented:
Fixes #171

jeffeaton commented:
Thanks, this looks good.  I've just been using it to add a new report. Two questions so far:

* Why is `use_*(..., locate = FALSE)` default? It would be nice for `locate = TRUE` so that this would 'just work':

```
> library(orderly)
> orderly_use_package("naomi")
Error in orderly_config_get(root, locate) : Invalid input
```
* Then, similar to comment on #169, it would be nice for this to work before I have specified my artefact: 
```
> orderly_use_package("naomi", locate = TRUE)
Error in recipe_read_check_artefacts(info$artefacts, filename, path) : 
  At least one artefact required
```

and possibly before I have specified my script (after updating `orderly.yml` to have an artefact): 
```
> orderly_use_package("naomi", locate = TRUE, edit = FALSE)
Error: '/Users/jeff/Dropbox/Documents/Code/R/naomi-orderly-example/src/output-data-pack/orderly.yml:script' must be a scalar
```

I suspect that these are better to be addressed under a separate issue though.

* Then when I have an `orderly.yml` with a script and an artefact, I am getting an error: 
```
> orderly_use_package("naomi", locate = TRUE)
Error in tmp[!grepl("^\\s*(#.*)?$", tmp)][[1L]] : subscript out of bounds
```
The reason seems to be because I have `packages:` but nothing following it in my `orderly.yml`, which is probably invalid?:

```
script: datapack-export.R

artefacts:
  - data:
      description: PEPFAR Data Pack export file
      filenames: naomi_datapack.csv

packages:
```
This doesn't seem a super crazy thing for a user to have done -- if there's a way to either allow this, or give a more descriptive error message for why it has gone wrong.

-----
After I removed the `packages:` from `ordely.yml` to be;
```
script: datapack-export.R

artefacts:
  - data:
      description: PEPFAR Data Pack export file
      filenames: naomi_datapack.csv
```
it goes beautifully:
```
> orderly_use_package("naomi", locate = TRUE)
Changes to 'src/output-data-pack/orderly.yml'
6 |       filenames: naomi_datapack.csv
7 | 
8 | packages:
9 |   - naomi
Write changes to file?  

1: no
2: yes

Selection: 2
Writing to 'src/output-data-pack/orderly.yml'
```
The prompt is a very nice feature.

richfitz commented:
Thanks Jeff - very useful.  Comments below (quoting an abbreviated version of your comment for context)

> * Why is `use_*(..., locate = FALSE)` default? It would be nice for `locate = TRUE` so that this would 'just work':

Fixed now in this PR

> * Then, similar to comment on #169, it would be nice for this to work before I > have specified my artefact: 
> ```
> > orderly_use_package("naomi", locate = TRUE)
> Error in recipe_read_check_artefacts(info$artefacts, filename, path) : 
>   At least one artefact required
> ```
> 
> and possibly before I have specified my script (after updating `orderly.yml` to > have an artefact): 
> ```
> > orderly_use_package("naomi", locate = TRUE, edit = FALSE)
> Error: '/Users/jeff/Dropbox/Documents/Code/R/naomi-orderly-example/src/output-data-pack/orderly.yml:script' must be a scalar
> ```

Fixed in #176, VIMC-3510 (you reported this also in #169)

> * Then when I have an `orderly.yml` with a script and an artefact, I am getting > an error: 
> ```
> > orderly_use_package("naomi", locate = TRUE)
> Error in tmp[!grepl("^\\s*(#.*)?$", tmp)][[1L]] : subscript out of bounds
> ```
> The reason seems to be because I have `packages:` but nothing following it in my `orderly.yml`, which is probably invalid?:

> ```
> script: datapack-export.R
> 
> artefacts:
>   - data:
>       description: PEPFAR Data Pack export file
>       filenames: naomi_datapack.csv
> 
> packages:
> ```
>
> This doesn't seem a super crazy thing for a user to have done -- if there's a way to either allow this, or give a more descriptive error message for why it has gone wrong.

I can't replicate this - could you please open another ticket with full details to reproduce.  Adding a line `packages:` to the example that is produced by `orderly::orderly_example("minimal")` I get:

```
> writeLines(readLines("orderly.yml"))
data:
  dat:
    query: SELECT name, number FROM thing
script: script.R
artefacts:
  staticgraph:
    description: A graph of things
    filenames: mygraph.png
packages:
> orderly::orderly_develop_start()
[ name       ]  example
[ data       ]  source => dat: 20 x 2
```

jeffeaton commented:
Thanks very much. New issue for the last created in #177.

r-ash commented on tests/testthat/test-util.R:
```diff
@@ -698,3 +698,93 @@ test_that("random_seed", {
+  writeLines(text, path)
+
+  expect_message(
+    str <- capture.output(
```
This is a new thing for me and shows up something I haven't seen before. When I've done things like this where I want to write tests about the messages printed and the result returned I've used `testthat::evaluate_promise` which I think is doing something similar to here but it returns the `result`, `message` and the `output`. So I wasn't aware the was a difference here.

What is the difference between `output` and `result`? Result is the return from the function and the `output` is what is coming from call to `cat`. Perhaps I don't quite get how `cat` works. How is `cat` different from a `message` or from `print`?

richfitz replied:
I was looking for `testthat::evaluate_promise` but could not find it - I'll rewrite to use that as it will be clearer.

cat writes to stdout, message to stderr, print *probably* writes to stdout, but it relies on underlying cat/message calls to do the printing

r-ash commented on R/util.R:
```diff
@@ -730,3 +730,96 @@ random_seed <- function(envir = globalenv()) {
+    return(list(name = name, exists = FALSE, block = FALSE))
+  }
+
+  re <- sprintf("^%s\\s*:\\s*(#.*)?", name)
```
Oh I see that line lower down has been removed since I wrote this - feel free to ignore.

richfitz commented:
Thanks @r-ash - just made that change now. Thanks for the reminder of that function, I really couldn't remember what it was called...
